### DISCUSSION TITLE
Filtering Custom Queries for MSSQL

### DISCUSSION THREAD

**Discussion by mintonan** [Budibase/budibase#11357] [Help] [answered]
Hi All,

I am running into issues with trying to filter a custom query that has been imported into Budibase.

![image](https://github.com/Budibase/budibase/assets/124909937/65d71e55-6a45-420b-9270-5926cce3d932)

In this image above I have two filtering options for start and end date, this is set to the infusion date column, as the users will want to filter out in-between two specific dates to find the data. I noticed though that when I am trying to use the front end options for filtering I do not have any options to put anything > or < for filtering,

Is this something that someone has looked into before? If I need to use some Javascript for this I am wondering if anyone has already made something for filtering dates so I don't need to reinvent the wheel here.

**Comment by melohagan** [comment]
![Screenshot 2023-08-01 at 16 31 18](https://github.com/Budibase/budibase/assets/101575380/34f1aa83-023b-4a9f-bb4e-f013e57f03a4)

Hey @mintonan Are the *Start Date* and *End Date* columns definitely [Date](https://docs.budibase.com/docs/datetime) types?

If they are text fields then you won't have the more than / less than filter options.

**Reply by mintonan** [reply]
Hi @melohagan Thanks for responding! 

Yes the column infusiondate is of type DateTime.

Here is a little bit of extra context as well to hopefully clarify any points

In the custom query I have made I have set two bindings together one called StartDate and one Called EndDate. If you take a look at the screenshot below they are both matched to the InfusionDate column. And when I run this it works fine and pulls all the data needed.
![image](https://github.com/Budibase/budibase/assets/124909937/c897786e-aa9c-4adb-8edc-9108bcec8b77)

What I was hoping to achieve was that with Budibase I could filter these options by changing the binding from the default value I have given to the value equal to the text entry boxes based off of my previous screenshot. But I cannot seem to find that option.

I have seen that there is a daterange component that have some options but my users would prefer to choose between two specific dates. 

If I am missing something please let me know, thanks for the help! :)

**Reply by melohagan** [reply]
Thanks for the extra info @mintonan 

Scroll down to the **Results** section after running your query, then click on the `Schema` tab.

Find your date fields and change the type to `Date`. Then make sure to click the `Save Query` button! This will ensure that the custom query will correctly use Date types instead of Text.

For example:
![Screenshot 2023-08-01 at 17 11 43](https://github.com/Budibase/budibase/assets/101575380/462defa3-56ab-4848-acb4-c909b29170da)

**Reply by mintonan** [reply]
Huh, weirdly enough after I did this the options for More than or equal to/Less than or equal to are still not there even after changing the column to date time.. Not sure why. I did save the query.

![image](https://github.com/Budibase/budibase/assets/124909937/361ba297-862d-48fa-9450-c42a9948e166)

![image](https://github.com/Budibase/budibase/assets/124909937/ac4effcd-7947-496c-a196-f9aa78ec11a9)

Is there something else I am missing to get those options you had in the first screenshot?

**Reply by melohagan** [reply]
Hey @mintonan I think the Design section hasn't refreshed the available filters, I was able to see the same thing too. Under the column name dropdown, I selected a different field, and then selected the original field again (in your case back to InfusionDate) to get the filter options to update.

I've raised a bug here: https://github.com/Budibase/budibase/issues/11420

**Reply by mintonan** [reply]
Yep! That worked, I think thats everything that I needed for now! Thanks for your help!

**Reply by mintonan** [reply]
Hey @melohagan I figured I would reply back here. (If you want me to start a new discussion I can do that as well) so after this was implemented I am now able to export this data with the correct filtering and it works great! But for some reason when I am putting the filters in, it is not updating the table itself on budibase.

Here is a screenshot showing what I mean, right now I have the filter going between August 1st 2021 to August 1st 2022 and on other screens I have made it usually will update the content I am filtering on the data provider, but here it isn't doing that.
![image](https://github.com/Budibase/budibase/assets/124909937/b60d1908-5fcc-4a69-9a59-0ae371645d54)

**Reply by melohagan** [reply]
Hey @mintonan 

Could you share a screenshot of what your bindings look like. So in your filter, how are you referencing the date form fields?

**Reply by mintonan** [reply]
![image](https://github.com/Budibase/budibase/assets/124909937/fb61fd84-85a0-449e-bb88-9f0c491d71d9)

StartInfusionDate is the field I used for this screenshot
![image](https://github.com/Budibase/budibase/assets/124909937/44f606ae-326d-420d-a3cc-15c30cd0fef5)

EndInfusionDate is this screenshot. I grabbed these from the menu on the right
![image](https://github.com/Budibase/budibase/assets/124909937/fe3da4f6-972d-4002-a263-935d8667d745)

Thanks again for the help!

**Reply by melohagan** [reply]
Hey @mintonan 

At a glance, I'm thinking that the more than / less than filters are expecting a number.

You could try the following JavaScript bindings instead:

```js
return new Date($("Audit Bag Form.Fields.EndFusionDate"))
```

or if that doesn't work maybe:

```js
return new Date($("Audit Bag Form.Fields.EndFusionDate")).getTime()
```

**Reply by mintonan** [reply]
Hey @melohagan  I did try this solution and for some reason its not working, 

I have noticed that even when I am just setting a datetime value for this it still does not work
![image](https://github.com/Budibase/budibase/assets/124909937/1a4bc9ad-252d-4668-89c8-0545c2723395)

InfusionDate should be after August 2021 after and before August 2022 before saving but here there are still years before that 
![image](https://github.com/Budibase/budibase/assets/124909937/c32cca26-27a3-43ea-97fe-8a50c94f5e2e)

Is there anything that needs to setup on the table itself I am missing?

**Comment by melohagan** [comment, marked as answer]
Hey @mintonan 

I'll dig into it and get back to you. I have a feeling it could be related to a recent bug fix I looked at involving custom queries for MySQL: https://github.com/Budibase/budibase/pull/11435

It may be the case that some type casting is not occurring as required.

As it is a custom query, you could always write the less than / more than into the query itself, and then pass in the binding values

**Reply by mintonan** [reply]
Hey @melohagan Thanks for the reply! 

Is there any documentation on how to do that? I have now set up a custom query with a more than/ less than but am not sure how to implement the changes for that

**Reply by melohagan** [reply]
There's some docs here on how to pass in the bindings: https://docs.budibase.com/docs/data-sources#custom-queries

**Reply by mintonan** [reply]
Yep that helped with the updates on the table! Thanks so much for the help!

**Reply by melohagan** [reply]
You're welcome! 😄

**Reply by melohagan** [reply]
Bug raised: https://github.com/Budibase/budibase/issues/11535
